### PR TITLE
Add the ability to select a whole command (or its output)

### LINKED ISSUE #4588: UI to Page / up and down by command + copy all output
It would be good if there was a UI element somewhere to page up or down to the previous or next command: [▲][▼]

This is something so useful, I'm not sure why no terminal emulator has it.

Would need a keybinding,  ctrl pageup / down ?

Similarly, useful:
Right click to copy all the output from a single command to the clipboard.

The terminal would probably need to have some idea about when it is rendering a prompt vs something else to enable this.

### PR DESCRIPTION
**Pull request by zadjii-msft**
Adds two new commands, `selectOutput` and `selectCommand`. These don't do much without shell integration enabled, unfortunately. If you do enable it, however, you can use these commands to quickly navigate the history to select whole commands (or their output).

Some sample JSON:

```json
        { "keys": "ctrl+shift+<", "command": { "action": "selectCommand", "direction": "prev" } },
        { "keys": "ctrl+shift+>", "command": { "action": "selectCommand", "direction": "next" } },
        { "keys": "ctrl+shift+[", "command": { "action": "selectOutput", "direction": "prev" } },
        { "keys": "ctrl+shift+]", "command": { "action": "selectOutput", "direction": "next" } },
```

**Demo gifs** in https://github.com/microsoft/terminal/issues/4588#issuecomment-1352042789 

closes #4588

Tested manually. 

<details>
<summary>CMD.exe user? It's dangerous to go alone! Take this.</summary>

Surely, there's a simpler way to do it, this is adapted from my own script.

```cmd
prompt $e]133;D$e\$e]133;A$e\$e\$e]9;9;$P$e\$e[30;107m[$T]$e[97;46m$g$P$e[36;49m$g$e[0m$e[K$_$e[0m$e[94m%username%$e[0m@$e[32m%computername%$e[0m$G$e]133;B$e\
```

</details>

### CODE REVIEW

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 1]
```diff
@@ -2093,6 +2094,106 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         }
     }
 
+    void ControlCore::SelectCommand(const bool goUp)
+    {
+        const til::point start = HasSelection() ? (goUp ? _terminal->GetSelectionAnchor() : _terminal->GetSelectionEnd()) :
+                                                  _terminal->GetTextBuffer().GetCursor().GetPosition();
+        std::optional<DispatchTypes::ScrollMark> nearest{ std::nullopt };
+        const auto marks{ std::ranges::reverse_view(_terminal->GetScrollMarks()) };
+        for (auto&& m : marks)
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/ControlCore.cpp:2103]:
```suggestion
        for (const auto&& m : marks)
```

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 2]
```diff
@@ -2093,6 +2094,106 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+                    // stash this as the new match
+                    nearest = m;
+                }
+            }
+        }
+
+        if (nearest.has_value())
+        {
+            auto start = nearest->end;
+            auto end = *nearest->commandEnd;
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/ControlCore.cpp:2137]:
```suggestion
            const auto start = nearest->end;
            const auto end = *nearest->commandEnd;
```

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 3]
```diff
@@ -2093,6 +2094,106 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+        }
+    }
+
+    void ControlCore::SelectOutput(const bool goUp)
+    {
+        const til::point start = HasSelection() ? (goUp ? _terminal->GetSelectionAnchor() : _terminal->GetSelectionEnd()) :
+                                                  _terminal->GetTextBuffer().GetCursor().GetPosition();
+        std::optional<DispatchTypes::ScrollMark> nearest{ std::nullopt };
+        const auto marks{ std::ranges::reverse_view(_terminal->GetScrollMarks()) };
+        for (auto&& m : marks)
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/ControlCore.cpp:2154]:
```suggestion
        for (const auto&& m : marks)
```

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 4]
```diff
@@ -2093,6 +2094,106 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+            // command, OR it does but thinks that it was empty, then just skip
+            // it.
+            if (!m.commandEnd.has_value() || *m.commandEnd == m.end)
+            {
+                continue;
+            }
+            // If this mark is before/after the start of our search in the
+            // buffer, ...
+
+            auto inTheRightDirection = (goUp && (m.commandEnd < start)) || // prev
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/ControlCore.cpp:2115]:
```suggestion
            const auto inTheRightDirection = (goUp && (m.commandEnd < start)) || // prev
```

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 5]
```diff
@@ -2093,6 +2094,106 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+        {
+            // If this mark doesn't know anything about the position of its
+            // output, OR it does but thinks that it was empty, then just skip
+            // it.
+            if (!m.outputEnd.has_value() || *m.outputEnd == *m.commandEnd)
+            {
+                continue;
+            }
+            // If this mark is before/after the start of our search in the buffer, ...
+            auto inTheRightDirection = (goUp && (m.outputEnd < start)) || // prev
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/ControlCore.cpp:2164]:
```suggestion
            const auto inTheRightDirection = (goUp && (m.outputEnd < start)) || // prev
```

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 6]
```diff
@@ -2093,6 +2094,106 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+                    // stash this as the new match
+                    nearest = m;
+                }
+            }
+        }
+
+        if (nearest.has_value())
+        {
+            auto start = *nearest->commandEnd;
+            auto end = *nearest->outputEnd;
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/ControlCore.cpp:2186]:
```suggestion
            const auto start = *nearest->commandEnd;
            const auto end = *nearest->outputEnd;
```

File: src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp [hunk 7]
```diff
@@ -402,6 +405,8 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
                 { ShortcutAction::Quit, RS_(L"QuitCommandKey") },
                 { ShortcutAction::AdjustOpacity, L"" }, // Intentionally omitted, must be generated by GenerateName
                 { ShortcutAction::RestoreLastClosed, RS_(L"RestoreLastClosedCommandKey") },
+                { ShortcutAction::SelectCommand, MustGenerate },
+                { ShortcutAction::SelectOutput, MustGenerate },
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:409]: I mean, I'm down to use `MustGenerate`, but you might as well replace the other `L""` while you're here, right?

File: src/cascadia/TerminalSettingsModel/Resources/en-US/Resources.resw [hunk 8]
```diff
@@ -661,4 +661,16 @@
...
+  </data>
+  <data name="SelectOutputPreviousCommandKey" xml:space="preserve">
+    <value>Select previous command output</value>
+  </data>
+  <data name="SelectCommandNextCommandKey" xml:space="preserve">
+    <value>Select next command</value>
+  </data>
+  <data name="SelectCommandPreviousCommandKey" xml:space="preserve">
+    <value>Select previous command</value>
+  </data>
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsModel/Resources/en-US/Resources.resw:675]:
```suggestion
  <data name="SelectCommandNextCommandKey" xml:space="preserve">
    <value>Select next prompt</value>
  </data>
  <data name="SelectCommandPreviousCommandKey" xml:space="preserve">
    <value>Select previous prompt</value>
  </data>
```

Would that be more clear? Should we rename the action to `selectPrompt`?

**Review by carlos-zamora — APPROVED** [review]
Just nits, honestly

File: src/cascadia/TerminalSettingsModel/Resources/en-US/Resources.resw (hunk 8 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalSettingsModel/Resources/en-US/Resources.resw:675]: I'm gonna hard NAK this - the "prompt" is the `C:\Users\migrie>` part. The "command" is `ping 8.8.8.8`, and the output is {you get it}.

I didn't think there was value in just selecting the prompts, but commands and output made sense.

File: src/cascadia/TerminalControl/ControlCore.cpp (hunk 1 above)
Comment (lhecker) [reply src/cascadia/TerminalControl/ControlCore.cpp:2103]: `auto&&` already does this. The `T&&` is meant for moving data, but because C++ was designed by people with good humor it's completely backwards compatible to all the other brilliant design decisions in the language. So there's 2 really funny things at play here:

"move-references" (aka `&&`) don't actually exist in the language. They're regular references (`&`) and `&&` only exists to select the proper overload at compile time. So if you have a `Foo(auto&)` and `Foo(auto&&)` and you call it with `Foo(std::move(foo))` it'll select the latter. But the _type of the parameter_ once you're inside the `Foo` function will NOT be `&&` anymore - it'll be `&` now! This is the sole reason why if you want to forward a to-be-moved type you have to move it again (hilarious, I know). This is wrong and makes a copy of the string:
```cpp
void Foo::Foo(std::string&& str) : _str{ str } {}
```
This is correct:
```cpp
void Foo::Foo(std::string&& str) : _str{ std::move(str) } {}
```

The second thing builds on top of the first: `&&` is a complete lie. Sometimes.

C++ was designed with humor instead of consistency in mind, considerate of not ever breaking the language, so if you write `T&&` anywhere and `T` is _not_ generic you get "move-references". In other words this can't be called with a `const std::string` (as you surely already noticed):
```cpp
void foo(std::string&& value) { ... }
```

But this isn't true for generic types (like `auto` or regular templates). Imagine you write a function like this and you call it with a `std::string` variable as the argument. These are the types you'd get out of it:
```cpp
template<typename T>
void foo(const T& value) {
    T               // std::string
    decltype(value) // const std::string&
}
```
Now what if the parameter is a `&&` and pass it the same variable (i.e. without `std::move`)?
```cpp
void foo(T&& value) {
    T               // std::string&    <--- !!!
    decltype(value) // std::string&
}
```
`T` is no the base type anymore! This is because (in plain English) the `&&` tells the compiler not just that this is a reference parameter, but that it should make the type `T` fully generic, including over whether it's a reference, const, volatile, etc. This is unlike the previous example where `T` will resolve to the base type! You can even call the above generic function with a `const std::string`. The type of `T` will then simply be `const std::string&`.

This results in `auto&&` being able to accept a `const T&` and this is exactly what the type of `m` will be if `marks` is `const` (which it is).

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 9]
```diff
@@ -12,6 +12,7 @@
 #include <unicode.hpp>
 #include <WinUser.h>
 #include <LibraryResources.h>
+#include <ranges>
```
Comment (lhecker) [inline src/cascadia/TerminalControl/ControlCore.cpp:15]: I'm not 100% sure if we should use `<ranges>` just yet. I've heard it can have a pretty significant compile duration impact and some testing indicates that this one use costs about 400ms on my machine (0.1%). Not that this matter to us, but to put this into perspective for some fun about generic languages in general, I can build the entirety of SQLite in about the same time lmao (245k lines of C).

More importantly however, I'd prefer if ranges was used consistently throughout the entire project instead of this one use. It'd be a good 1h Area-CodeHealth fun to replace all of our reverse-iterator uses and complex find()-loops with ranges all at once. This would also help figure out what the binary size impact will be (which I've heard is even more significant than the compile time impact).

Also, instead of `std::ranges::reverse_view{}` (which is the type) you should use `std::views::reverse()` (the factory). Basically either of these are the usually expected usage:
```cpp
const auto& marks = _terminal->GetScrollMarks();
for (const auto& m : std::views::reverse(marks)) {}
for (const auto& m : marks | std::views::reverse) {}
```

Alternatively a regular reverse-iterator loops works as well of course:
```cpp
auto it = marks.rbegin();
const auto end = marks.rend();
for (; it != end; ++it)
{
    const auto& m = *it;
}
```

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 10]
```diff
@@ -2093,6 +2094,106 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+            if (inTheRightDirection)
+            {
+                // ... and we either haven't found a match, or the current nearest
+                // is after/before this mark in the buffer
+                if (!nearest.has_value() ||
+                    ((goUp && (*m.commandEnd > *nearest->commandEnd)) || // prev
+                     (!goUp && (m.end < *nearest->commandEnd)))) // next
+                {
+                    // stash this as the new match
+                    nearest = m;
```
Comment (lhecker) [inline src/cascadia/TerminalControl/ControlCore.cpp:2129]: I don't entirely get why we use a reverse-iteration if there's no early-exit condition anyways. Can't this just forward-iterate then?

Also, instead of using `inTheRightDirection` can't you just calculate a signed distance value for `start - m.commandEnd` or `m.end - start`? If you track the best distance and don't throw it away you can drop the check here and merge it with `inTheRightDirection`. It'd be just this as far as I can see:
```cpp
// Early return so we don't have to check for the validity of `nearest` below after the loop exits.
if (marks.empty() {
    return;
}

// `CoordTypeMax` doesn't exist yet. See here:
//   https://github.com/microsoft/terminal/blob/db90593275091edd6a7ce07a37be89fc15fa5a45/src/inc/til/point.h#L9-L10
static constexpr til::point worst{ til::CoordTypeMax, til::CoordTypeMax };

DispatchTypes::ScrollMark nearest;
auto bestDistance = worst;

for (const auto& m : marks) {
    const auto distance = goUp
        ? start - m.commandEnd.value_or(worst)
        : m.end - start;
    if (dinstance && distance < bestDistance) {
        nearest = m;
    }
}
```

Same for the other function below.

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 11]
```diff
@@ -2093,6 +2094,106 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+        const til::point start = HasSelection() ? (goUp ? _terminal->GetSelectionAnchor() : _terminal->GetSelectionEnd()) :
+                                                  _terminal->GetTextBuffer().GetCursor().GetPosition();
+        std::optional<DispatchTypes::ScrollMark> nearest{ std::nullopt };
+        const auto marks{ std::ranges::reverse_view(_terminal->GetScrollMarks()) };
+        for (auto&& m : marks)
+        {
+            // If this mark doesn't know anything about the position of its
+            // command, OR it does but thinks that it was empty, then just skip
+            // it.
+            if (!m.commandEnd.has_value() || *m.commandEnd == m.end)
```
Comment (lhecker) [inline src/cascadia/TerminalControl/ControlCore.cpp:2108]: This condition makes me think that `ScrollMark` could use some small helper member functions. If I'm reading this right, this would be something like `!m.HasCommand()`? Because with that in place we can also drop the comment here.

**Review by lhecker — APPROVED** [review]

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 12]
```diff
@@ -2182,6 +2182,90 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+            const auto bufferSize{ _terminal->GetTextBuffer().GetSize() };
+            bufferSize.DecrementInBounds(end);
+
+            auto lock = _terminal->LockForWriting();
+            _terminal->SelectNewRegion(start, end);
+            _renderer->TriggerSelection();
+        }
+    }
+
+    void ControlCore::SelectOutput(const bool goUp)
```
Comment (DHowett) [inline src/cascadia/TerminalControl/ControlCore.cpp:2230]: these seem to be the same function with a one-line difference

File: src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp [hunk 13]
```diff
@@ -329,22 +331,24 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
...
+        // Sentinel used to indicate this command must ALWAYS be generated by GenerateName
+        static const winrt::hstring MustGenerate{ L"" };
         // Use a magic static to initialize this map, because we won't be able
         // to load the resources at _init_, only at runtime.
         static const auto GeneratedActionNames = []() {
             return std::unordered_map<ShortcutAction, winrt::hstring>{
                 { ShortcutAction::AdjustFontSize, RS_(L"AdjustFontSizeCommandKey") },
                 { ShortcutAction::CloseOtherPanes, RS_(L"CloseOtherPanesCommandKey") },
-                { ShortcutAction::CloseOtherTabs, L"" }, // Intentionally omitted, must be generated by GenerateName
+                { ShortcutAction::CloseOtherTabs, MustGenerate },
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:342]: clever
